### ISSUE TITLE
Function gather() implementation

### ISSUE THREAD

**Issue by zhulf0804** [HuguesTHOMAS/KPConv-PyTorch#132]
Hi HuguesTHOMAS,

Thanks for your great work. I have two questions about the function `gather()` implemented in [https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/models/blocks.py#L35](https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/master/models/blocks.py#L35). I find it's really much faster using `method=2` compared to `method=0`. And the questions are:

- Why `method = 2` is faster than `method = 0` in the function `gather()`?
- `method = 2` will also bring randomness in the training backwards process. Do you know why is there randomness in the `method = 2` implementation ?

Looking forward to your reply.

**Comment by HuguesTHOMAS** [comment]
Hi @zhulf0804,

It has been a while since I implemented this function. If I remember correctly, I think it was because of the way the backward gradient propagation is implemented for the operation:
```python
x[indices]
```
The function `gather` is much better optimized. I don't remember the sources I found for this, but a quick google search should give you more details.

> method = 2 will also bring randomness in the training backwards process

Why do you say that? I don't recall that there is any random operation in method 2. Could you elaborate?

Best,
Hugues

**Comment by zhulf0804** [comment]
Hi HuguesTHOMAS, 

Thanks for your reply. I found the randomness when i re-implement [PREDATOR](https://overlappredator.github.io) (a point cloud registration network) [here](https://github.com/zhulf0804/PREDATOR) based on [KPConv](https://github.com/HuguesTHOMAS/KPConv-PyTorch).  

About training:
I opened multiple terminals to train the model at the same time and i print the loss value in each iteration during training. The loss value is a bit different between multiple terminals, except the loss after the first iteration (which has not been back propagated).

After updating `method = 2` with `method = `0`, the loss value is fixed when i opened multiple terminals to train.

About test:
The loss value is fixed when i opened multiple terminals to test, even with `method = 2`. So it doesn't affect the forward pass.

Some days later, a friend sent me a link [https://github.com/pytorch/pytorch/issues/15245](https://github.com/pytorch/pytorch/issues/15245), maybe there is a small bug  in PyTorch.
